src/js/background/sync.js

```diff
@@ -0,0 +1,461 @@
+      await browser.contextualIdentities.onRemoved.hasListener(listenerList[1]) &&
+      await browser.contextualIdentities.onUpdated.hasListener(listenerList[2])
+    );
+  }
+
+};
+
+sync.init();
+
+async function restore() {
+  console.log("restore");
+  await reconcileIdentities();
+  await reconcileSiteAssignments();
+  return;
+}
+
+/*
+ * Checks for the container name. If it exists, they are assumed to be the
+ * same container, and the color and icon are overwritten from sync, if
+ * different.
+ */
+async function reconcileIdentities(){
+  console.log("reconcileIdentities");
+
+  // first delete any from the deleted list
+  const deletedIdentityList =
+    await sync.storageArea.getDeletedIdentityList();
+  // first remove any deleted identities
+  for (const deletedUUID of deletedIdentityList) {
+    const deletedCookieStoreId = 
+      await identityState.lookupCookieStoreId(deletedUUID);
+    if (deletedCookieStoreId){
+      try{
+        await browser.contextualIdentities.remove(deletedCookieStoreId);
+      } catch (error) {
+        // if the identity we are deleting is not there, that's fine.
+        console.error("Error deleting contextualIdentity", deletedCookieStoreId);
+        continue;
+      }
+    }
+  }
+
+  const localIdentities = await browser.contextualIdentities.query({});
+  const syncIdentities = 
+    await sync.storageArea.getIdentities();
+  const cookieStoreIDmap = 
+    await sync.storageArea.getCookieStoreIDMap();
+  // now compare all containers for matching names.
+  for (const syncIdentity of syncIdentities) {
+    syncIdentity.macAddonUUID = cookieStoreIDmap[syncIdentity.cookieStoreId];
+    if (syncIdentity.macAddonUUID){
+      const localMatch = localIdentities.find(
+        localIdentity => localIdentity.name === syncIdentity.name
+      );
+      if (!localMatch) {
+        // if there's no name match found, check on uuid,
+        const localCookieStoreID = 
+          await identityState.lookupCookieStoreId(syncIdentity.macAddonUUID);
+        if (localCookieStoreID) {
+          await ifUUIDMatch(syncIdentity, localCookieStoreID);
+          continue;
+        }
+        await ifNoMatch(syncIdentity);
+        continue;
+      }
+      await ifNamesMatch(syncIdentity, localMatch);
+      continue;
+    }
+    // if no macAddonUUID, there is a problem with the sync info and it needs to be ignored.
+  }
+
+  await updateSiteAssignmentUUIDs();
+
+  async function updateSiteAssignmentUUIDs(){
+    const sites = assignManager.storageArea.getAssignedSites();
+    for (const siteKey of Object.keys(sites)) {
+      await assignManager.storageArea.set(siteKey, sites[siteKey]);
+    }
+  }
+}
+
+async function ifNamesMatch(syncIdentity, localMatch) {
+  // Sync is truth. if there is a match, compare data and update as needed
+  if (syncIdentity.color !== localMatch.color 
+      || syncIdentity.icon !== localMatch.icon) {
+    await browser.contextualIdentities.update(
```

Comment:
The name of this method should match it's functionality as this is a destructive action.

Reply:
I just realised the naming here, they should be named as if it could be used outside from the current callsite.